iceoryx_posh/include/iceoryx_posh/popo/trigger_state.hpp

```diff
@@ -0,0 +1,95 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_POSH_POPO_TRIGGER_STATE_HPP
+#define IOX_POSH_POPO_TRIGGER_STATE_HPP
+
+#include "iceoryx_posh/internal/log/posh_logging.hpp"
+#include "iceoryx_utils/cxx/function_ref.hpp"
+#include "iceoryx_utils/error_handling/error_handling.hpp"
+
+#include <cstdint>
+#include <limits>
+
+namespace iox
+{
+namespace popo
+{
+/// @brief TriggerState holds the state of a trigger and is the base class
+///        for trigger.
+class TriggerState
+{
+  public:
+    static constexpr uint64_t INVALID_TRIGGER_ID = std::numeric_limits<uint64_t>::max();
+    template <typename T>
+    using Callback = void (*)(T* const);
+
+    /// @brief constructs an empty TriggerState
+    TriggerState() = default;
+    virtual ~TriggerState() = default;
+
+    /// @brief constructs a TriggerState object
+    /// @param[in] origin pointer to the origin of the TriggerState
+    /// @param[in] triggerId id of the trigger
+    /// @param[in] callback the callback of the trigger
+    template <typename T>
+    TriggerState(T* const origin, const uint64_t triggerId, const Callback<T> callback) noexcept;
+
+    /// @brief returns the trigger id
+    /// @return the empty TriggerState always returns INVALID_TRIGGER_ID, otherwise the actual triggerId is returned
+    /// which can also be INVALID_TRIGGER_ID
+    uint64_t getTriggerId() const noexcept;
+
+    /// @brief confirms the origin
+    /// @param[in] origin the possible origin
+    /// @return true if the address is equal to the origin, otherwise false. The empty TriggerState returns always
+    /// false.
+    template <typename T>
+    bool doesOriginateFrom(T* const origin) const noexcept;
+
+    /// @brief returns the pointer to the origin.
+    /// @return If T equals the origin type it returns the pointer to the origin.
+    /// Otherwise it calls the errorHandler with a moderate error of
+    /// kPOPO__TRIGGER_STATE_TYPE_INCONSISTENCY_IN_GET_ORIGIN and returns nullptr.
+    template <typename T>
+    T* getOrigin() noexcept;
+
+    /// @brief returns the pointer to the origin.
+    /// @return If T equals the origin type it returns the pointer to the origin.
+    /// Otherwise it calls the errorHandler with a moderate error of
+    /// kPOPO__TRIGGER_STATE_TYPE_INCONSISTENCY_IN_GET_ORIGIN and returns nullptr.
+    template <typename T>
+    const T* getOrigin() const noexcept;
+
+    /// @brief If a callback is set it executes the callback.
+    /// @return true if the callback was called, otherwise false
+    bool operator()() const noexcept;
+
+    friend class Trigger;
+
+  protected:
+    void* m_origin = nullptr;
+    uint64_t m_originTypeHash = 0U;
+    uint64_t m_triggerId = INVALID_TRIGGER_ID;
+
+    Callback<void> m_callbackPtr = nullptr;
+    cxx::function_ref<void(void* const, Callback<void>)> m_callback;
```

Comment:
Why are some callbacks in `TriggerState` and others in `Trigger`? Didn't get the rationale behind it.

Reply:
The Trigger is contained inside of the WaitSet. The WaitSet needs to acquire the information if the trigger was triggered (hasTriggeredCallback) and needs to cleanup if it goes out of scope (resetCallback).

The TriggerState is acquired by the user when calling `WaitSet.wait()` and contains the triggerId, the user defined callback and the pointer to the origin. The user should never call the resetCallback or the hasTriggeredCallback and the WaitSet is never interested in the origin or the user callback.

Reply:
More documentation is needed. Maybe also think about the class names one more time. Is this really a state?

Reply:
@MatthiasKillat your 2 cents are perfect ... I will steal them for the documentation! :D